### PR TITLE
Add Android Documents Provider

### PR DESCRIPTION
**Pull request by brunotl**
allow access to "/Android/data/org.xcsoar/files" from android file explorer

fix issue #992

Brief summary of the changes
----------------------------

Related issues and discussions
------------------------------

### CODE REVIEW

**Comment by lordfolken** [comment]
hey bruno, 
i tried as per your suggestion, but was failing at the androidx dependency. But i see that you resolved that now.

**Comment by brunotl** [comment]
> hey bruno, i tried as per your suggestion, but was failing at the androidx dependency. But i see that you resolved that now.

androidx dependency was only for few java annotation, I've just removed the annotations ...

**Comment by MaxKellermann** [comment]
But ... this doesn't fix #992, does it? #992 is about `XCSoarData` not being accessible anymore to the XCSoar app, but your code does not make it accessible, therefore it's not a solution to #992.
Quite contrary, all it does is add a shortcut to reach into the new-style data folder from other apps.

And, correct me if I'm wrong (I don't know these APIs): it effectively removes all protection from XCSoar's data folders, allowing arbitrary untrusted apps to manipulate any of XCSoar's data files. If this is true, then this opens a rather critical vulnerability, and allows arbitrary untrusted apps to invade the user's privacy.

**Comment by brunotl** [comment]
> But ... this doesn't fix #992, does it? #992 is about `XCSoarData` not being accessible anymore to the XCSoar app, but your code does not make it accessible, therefore it's not a solution to #992. Quite contrary, all it does is add a shortcut to reach into the new-style data folder from other apps.

right...

> 
> And, correct me if I'm wrong (I don't know these APIs): it effectively removes all protection from XCSoar's data folders, allowing arbitrary untrusted apps to manipulate any of XCSoar's data files. If this is true, then this opens a rather critical vulnerability, and allows arbitrary untrusted apps to invade the user's privacy.

it does not remove any protection, it only allow user to manage the file using the Android UI. trusted (or untrusted) app can't use this provider without asking to the user to select file using the native android file picker...

**Comment by lordfolken** [comment]
Its true that its not a "technical solution" in the sense that you are able to reuse the XCSoarData on the "Sdcard". But you already ruled that out, as it would require to use google's high level apis that don't allow many of the operations allowed by current posix standard.

But the big problem right now is, that users cannot access their flight logs and that they cannot add custom files to xcsoar's data dir. The File manager per default blocks all direct access to this directory.

Its not "arbituary untrusted apps". They need to have the storage "manage all files" permission in order to access it. Otherwise they are locked into their application directory just like xcsoar is  locked into its. And if threy would, they would have to allow explicit access via File Manager.

File: android/src/AllFilesDocumentsProvider.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+
+        if (Document.MIME_TYPE_DIR.equals(mimeType)) {
+            if (!file.mkdirs()) {
+                throw new FileNotFoundException("Failed to create directory with name " +
+                        displayName + " and documentId " + parentDocumentId);
+            }
+        } else {
+            try {
+                file.createNewFile();
+                file.setWritable(true);
```
Comment (lordfolken) [inline android/src/AllFilesDocumentsProvider.java:310]: Should this be:
public boolean setWritable(boolean writable, boolean ownerOnly) 
so therefore true,true ? 
https://www.demo2s.com/android/android-file-setwritable-boolean-writable-boolean-owneronly.html

Comment (brunotl) [reply android/src/AllFilesDocumentsProvider.java:310]: it's : https://developer.android.com/reference/java/io/File#setWritable(boolean)

**Comment by MaxKellermann** [comment]
Amended to update the testing/AndroidManifest.xml and removed the "fixes" tag (because this PR does not fix #995 

1e45a795f0c762fd6e4194abda0cc2b9569d8553

**Pull request closed without merging**